File: cpp/leetcode1_twosum.cpp
#include<vector>
#include<unordered_map>
using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> seen;
        for (int i = 0; i < nums.size(); i++){
            int candidate = nums[i];
            int complement = target - candidate;

            if (seen.count(complement)){
                return {seen[complement], i};
            }

            seen[candidate] = i;
        }

        return {};
    }
};

File: python/leetcode1_twosum.py
from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        seen = {} # use hash map to get O(n)

        for i, num in enumerate(nums):
            complement = target - num

            if complement in seen:
                return [seen[complement], i]
            
            seen[num] = i